core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java

```diff
@@ -0,0 +1,420 @@
+        final boolean useLongPolling;
+        if (longPollingTimeoutMillis > 0) {
+            final String expectedState =
+                    Ascii.toLowerCase(req.headers().get(HttpHeaderNames.IF_NONE_MATCH, ""));
+            if ("\"healthy\"".equals(expectedState) || "w/\"healthy\"".equals(expectedState)) {
+                useLongPolling = isHealthy;
+            } else if ("\"unhealthy\"".equals(expectedState) || "w/\"unhealthy\"".equals(expectedState)) {
+                useLongPolling = !isHealthy;
+            } else {
+                useLongPolling = false;
+            }
+        } else {
+            useLongPolling = false;
+        }
+
+        final HttpMethod method = ctx.method();
+        if (useLongPolling) {
+            // Disallow other methods than HEAD/GET for long polling.
+            switch (method) {
+                case HEAD:
+                case GET:
+                    break;
+                default:
+                    throw HttpStatusException.of(HttpStatus.METHOD_NOT_ALLOWED);
+            }
+
+            assert healthCheckerListener != null : "healthCheckerListener is null.";
+            assert pendingHealthyResponses != null : "pendingHealthyResponses is null.";
+            assert pendingUnhealthyResponses != null : "pendingUnhealthyResponses is null.";
+
+            // If healthy, wait until it becomes unhealthy, and vice versa.
+            synchronized (healthCheckerListener) {
+                final boolean currentHealthiness = isHealthy();
+                if (isHealthy == currentHealthiness) {
+                    final CompletableFuture<HttpResponse> future = new CompletableFuture<>();
+                    final ScheduledFuture<Boolean> timeoutFuture = ctx.eventLoop().schedule(
+                            () -> future.complete(HttpResponse.of(notModifiedHeaders)),
+                            longPollingTimeoutMillis, TimeUnit.MILLISECONDS);
+                    final PendingResponse pendingResponse = new PendingResponse(method, future, timeoutFuture);
+                    if (isHealthy) {
+                        pendingUnhealthyResponses.add(pendingResponse);
+                    } else {
+                        pendingHealthyResponses.add(pendingResponse);
+                    }
+
+                    updateRequestTimeout(ctx, longPollingTimeoutMillis);
+                    return HttpResponse.from(future);
+                } else {
+                    // State has been changed before we acquire the lock.
+                    // Fall through because there's no need for long polling.
+                }
+            }
+        }
+
+        switch (method) {
+            case HEAD:
+            case GET:
+                return newResponse(method, isHealthy);
+            case CONNECT:
+            case DELETE:
+            case OPTIONS:
+            case TRACE:
+                return HttpResponse.of(HttpStatus.METHOD_NOT_ALLOWED);
+        }
+
```

Comment:
Think it helps to add a comment like `// POST or PUT`

Reply:
Added an assertion.